color/src/color.rs

```diff
@@ -0,0 +1,531 @@
+    type Output = AlphaColor<T>;
+
+    fn mul(self, rhs: AlphaColor<T>) -> Self::Output {
+        rhs * self
+    }
+}
+
+impl<T: Colorspace> core::ops::Div<f32> for AlphaColor<T> {
+    type Output = Self;
+
+    #[expect(
+        clippy::suspicious_arithmetic_impl,
+        reason = "somebody please teach clippy about multiplicative inverses"
+    )]
+    fn div(self, rhs: f32) -> Self {
+        self * rhs.recip()
+    }
+}
+
+impl<T: Colorspace> core::ops::Add for AlphaColor<T> {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self {
+        let x = self.components;
+        let y = rhs.components;
+        Self::new([x[0] + y[0], x[1] + y[1], x[2] + y[2], x[3] + y[3]])
+    }
+}
+
+impl<T: Colorspace> core::ops::Sub for AlphaColor<T> {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self {
+        let x = self.components;
+        let y = rhs.components;
+        Self::new([x[0] - y[0], x[1] - y[1], x[2] - y[2], x[3] - y[3]])
+    }
+}
+
+impl<T: Colorspace> core::ops::Mul<f32> for PremulColor<T> {
+    type Output = Self;
+
+    fn mul(self, rhs: f32) -> Self {
+        Self::new(self.components.map(|x| x * rhs))
+    }
+}
+
+impl<T: Colorspace> core::ops::Mul<PremulColor<T>> for f32 {
+    type Output = PremulColor<T>;
+
+    fn mul(self, rhs: PremulColor<T>) -> Self::Output {
+        rhs * self
+    }
+}
+
+impl<T: Colorspace> core::ops::Div<f32> for PremulColor<T> {
+    type Output = Self;
+
+    #[expect(
+        clippy::suspicious_arithmetic_impl,
+        reason = "somebody please teach clippy about multiplicative inverses"
+    )]
+    fn div(self, rhs: f32) -> Self {
+        self * rhs.recip()
+    }
+}
+
+impl<T: Colorspace> core::ops::Add for PremulColor<T> {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self {
+        let x = self.components;
+        let y = rhs.components;
+        Self::new([x[0] + y[0], x[1] + y[1], x[2] + y[2], x[3] + y[3]])
+    }
+}
+
+impl<T: Colorspace> core::ops::Sub for PremulColor<T> {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self {
+        let x = self.components;
+        let y = rhs.components;
+        Self::new([x[0] - y[0], x[1] - y[1], x[2] - y[2], x[3] - y[3]])
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{fixup_hue, HueDirection};
+
+    #[test]
+    fn test_hue_fixup() {
+        // Verify that the hue arc matches the spec for all hues specified
+        // within [0,360).
+        for h1 in [0.0, 10.0, 180.0, 190.0, 350.0] {
+            for h2 in [0.0, 10.0, 180.0, 190.0, 350.0] {
+                let dh = h2 - h1;
+                let mut fixed_h2 = h2;
+                fixup_hue(h1, &mut fixed_h2, HueDirection::Shorter);
+                let (mut spec_h1, mut spec_h2) = (h1, h2);
+                if dh > 180.0 {
+                    spec_h1 += 360.0;
+                } else if dh < -180.0 {
+                    spec_h2 += 360.0;
+                }
+                assert_eq!(fixed_h2 - h1, spec_h2 - spec_h1);
```

Comment:
If I were writing this test, I'd probably put these all in their own "anonymous" scope, so that there's no risk of forgetting to reset a variable by accident.